### DISCUSSION TITLE
Environment variables not passing through in CI after Turborepo 2.0 upgrade

### DISCUSSION THREAD

**Discussion by yashsway** [vercel/turborepo#8905] [Help]
### Summary

So after the change from this PR https://github.com/vercel/turbo/pull/8182, environment variables in CI are no longer passing through to the Turbo pipeline.

Here are some iterations of the `turbo.json` configuration I've tried:

1. 
```
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": [
    "**/.env.*",
    "tsconfig.json",
    "tsconfig.base.json",
    "tsconfig.lib.json"
  ],
  "globalEnv": ["CIRCLE_*", "VERCEL_*"],
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:stage": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:integration": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:prod": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
   // ... excluded rest for brevity
  }
}
```

2.
```
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": [
    "**/.env.*",
    "tsconfig.json",
    "tsconfig.base.json",
    "tsconfig.lib.json"
  ],
  "env": ["CIRCLE_*", "VERCEL_*"],
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:stage": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:integration": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:prod": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
   // ... excluded rest for brevity
  }
}
```

3.
```
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": [
    "**/.env.*",
    "tsconfig.json",
    "tsconfig.base.json",
    "tsconfig.lib.json"
  ],
  "tasks": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ]
    },
    "build:stage": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ],
      "passThroughEnv": ["CIRCLE_*", "VERCEL_*"]
    },
    "build:integration": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ],
      "passThroughEnv": ["CIRCLE_*", "VERCEL_*"]
    },
    "build:prod": {
      "dependsOn": ["^build"],
      "outputs": [
        "dist/**",
        "build/**",
        ".next/**",
        "!.next/cache/**",
        "lib/**"
      ],
      "passThroughEnv": ["CIRCLE_*", "VERCEL_*"]
    },
   // ... excluded rest for brevity
  }
}
```

None work. 😢 

And here's an iteration of the `package.json` `scripts` config I tried, that worked! 🪂 :
```
  "scripts": {
    "analyze": "turbo run analyze",
    "build": "turbo run build",
    "build:integration": "turbo run build:integration --env-mode=loose",
    "build:prod": "turbo run build:prod --env-mode=loose",
    "build:stage": "turbo run build:stage --env-mode=loose",
    // ... excluded rest for brevity
 }
```

There might be an issue with the functionality or the documentation here https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#environment-modes because the documentation says that `env` or `globalEnv` are considered but there's a section on passthrough variables https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#passthrough-variables that's not mentioned in the CI vendor compat section https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#ci-vendor-compatibility or anywhere else that both `env` and `globalEnv` are mentioned... which doesn't make any sense because that's what I need! The only place where the pass through config options are mentioned again are https://turbo.build/repo/docs/reference/run#strict. The docs are very inconsistent on this.

These variables don't actually affect the cache ... so I just need them to pass through. If they're accounted for in these pass through configs, then I expect that they fall under the `strict` mode checks too as https://turbo.build/repo/docs/reference/run#strict indicates?

Going through the escape hatch and setting `--env-mode=loose` can't be the only option can it?

**Comment by anthonyshew** [comment]
The third iteration you've provided appears to be what you're looking for, though the `build` task doesn't have passthroughs assigned like the rest. Was that purposeful?

I'm also noticing that [the CircleCI system variables](https://kapeli.com/cheat_sheets/CircleCI_Standard_Variables.docset/Contents/Resources/Documents/index) have a gotcha, that the `CIRCLECI` variable doesn't follow the rest of the `CIRCLE_*` convention. I'm mentioning that _just  in case_, but it may not be your situation.

If neither of those are the case:
> These variables don't actually affect the cache ... so I just need them to pass through. If they're accounted for in these pass through configs, then I expect that they fall under the strict mode checks too as https://turbo.build/repo/docs/reference/run#strict indicates?

This is correct and how passthroughs work. Our tests show that things are working as expected, but its possible there's a bug. We would need a reproduction to investigate.
